Move `float64` to `layouts_beta`

ccasin opened the pull request:
This does what it says on the tin.  Note that it needs the fix from #1793 - that PR should be merged first, and for review look at just the last commit of this PR.

The vast majority of changes are in tests.  There were many CRs in the `typing-layouts` tests indicating float versions of void tests should be added.  I've done that.  I haven't removed the void versions - void is an odd special case and it seems useful to continue testing all its interactions.  Many other CRs have been updated to reflect the current state of affairs.  In `typing-layouts-float64`, I adjusted things so that, where possible, rather than having separate versions of a file for each maturity level, we just have separate `TEST` commands.

As discussed in the unboxed ints PR, I think these tests are adding a lot of value as a sanity check during development, but they are very repetitive and I think it's probably not worth doing a rigorous review of their diffs.  But, I'll leave that up to the reviewer (could be @goldfirere, @antalsz , or @ncik-roberts).

Finally, I'll note that this PR is going to be a bit conflicty with other PRs in flight (particularly #1769). That's fine - I'll try to resolve the conflicts when one gets merged.  I do think it would be nice to get both in before the next compile roll (and maybe even the yet-to-be-posted float+immediate records PR, but that might be too ambitious).

goldfirere commented on ocaml/testsuite/tests/typing-layouts/modules_alpha.ml:
```diff
@@ -7,6 +7,7 @@ type t_any   : any
 type t_value : value
 type t_imm   : immediate
 type t_imm64 : immediate64
+type t_float64 : float64
```
As with basics_alpha, I think there should basically be no changes in this file.

goldfirere commented on ocaml/testsuite/tests/typing-layouts/modules_beta.ml:
```diff
@@ -412,4 +560,5 @@ module F : sig end -> sig end
 (* Test 8: [val]s must be representable *)
 
-(* CR layouts: Bring this test back from modules_alpha *)
+(* CR layouts v2.5: Bring a float64 version of this test over when we allow
```
I think with the change I've suggested above, this could be done now.

goldfirere reviewed (commented):
No (new) bugs discovered, but I do think there is a better story than this around testing. If that's hard to sort out, I'm OK with merging this and cleaning up the testing later (though I think it shouldn't be hard to sort out).

Excited to see this moving forward!

ccasin commented on ocaml/parsing/builtin_attributes.ml:
```diff
@@ -478,7 +478,9 @@ let layout ~legacy_immediate attrs =
-     | Any | Void | Float64 ->
+     | Float64 ->
+        check Language_extension.(is_at_least Layouts Beta)
+     | Any | Void ->
```
I'm confused - my impression was you explicitly left out any from this in the previous PR that moved the rest of `any` to beta because you encountered some issues.  Is that not right?

goldfirere replied:
I don't recall that. Maybe you're thinking about the fact that I want to delete this function entirely? Which is true, but it's easy enough to make this small improvement in the meantime. So the problem is likely originally my oversight -- sorry.

ccasin replied:
OK, I took a look at this and don't see any fundamental obstacles, but there are a bunch more test CRs to address when we do it, so I suggest it should be its own PR.

ccasin commented:
@goldfirere Regarding the tests, it looks like you have two questions:
- Why am I testing some files on both basics and alpha?  Well, we have alpha in the system, and it seems useful to check its behavior.  For example, if in some PR we accidentally make it so that alpha allows mixed blocks, or some other typing bug that allows too much, we'll only catch that if we have tests for alpha.  Not the most likely bugs, admittedly, but they are simple to test for and I don't really understand what not testing alpha buys us - the maintenance cost is zero where alpha and beta agree (since the files can be combined, then, as I've done here).  And the runtime cost is negligible relative to the test suite. 
- Why do some alpha files have new tests?  Because I added those tests to the beta versions to address CRs, and the design has always been that the alpha tests are a superset of the beta tests.

ccasin commented:
@goldfirere and I discussed the alpha/beta tests question.  We agree that keeping up the standard that the alpha tests are a superset of the beta tests has both modest benefit and modest cost.  I think the benefit outweighs the cost, and he doesn't.  We compromised on keeping the alpha tests on in cases where only one copy of the test is needed, since the costs are particularly low there, and reverting this PRs changes to other alpha test files (so `float64`'s behavior is only checked in the beta version of those tests).  Note this leaves those files in a state where there isn't a clear standard about what belongs in them, and some future cleanup might be desirable, but it doesn't seem like a huge deal.

I've rebased this onto main to pick up the merged version of #1793.  I've also added a commit to roll back the test changes, and fixed or responded to the remaining review comments.
